module5/lessons/Week1/ErrorHandling.md

```diff
@@ -0,0 +1,326 @@
+
+The 400 error codes here are good place to start for API errors! [HTTP Cats](https://http.cat/)
+</section>
+
+<aside class="instructor-notes" markdown="1">
+    <p><strong>Instructor Note</strong><br>
+    Examples of things students might say:
+    Server:
+    * If an API you are calling is down
+    * If the API changes so the request now gives an error instead of data
+    * If the database is down
+    * If the database changes the password
+    * Missing necessary environmental variable
+    * Package changes version and the old version you are using doesn't work anymore
+    Client:
+    * Syntax errors like missing an equals in an API request.
+    * If you try to write something of the wrong type to the database
+    * If you try to write something to the database without a required field
+    * If you try to loop through something that is null
+    * If you try to go to a route that doesn't exist
+    * If you are not authorized to access a page</p>
+</aside>
+
+
+### What happens when an error occurs?
+
+Without proper error handling, if your .NET app breaks this is what happens. 
+
+![.NET Error](/assets/images/module5/week1/Error.png)
+
+Can I get a show of hands, who has seen something like this before?
+
+<section class="call-to-action" markdown="1">
+With your partner discuss the following questions: 
+
+* Is a good user experience? Why or why Not? 
+* What would make this user experience better?
+</section>
+
+## Error Handling Techniques
+
+## Checking for Null
+
+One of the most common edge cases that can cause your app to crash is when something expects a value but instead is given null.
+
+#### In the View
+
+What happens if what's passed in for the Movie is Null? Currently, our code doesn't have any error handling and we get this messy looking error we just discussed.
+
+![.Unhandled Exception](/assets/images/module5/week1/UnhandledException.png)
+
+If some value could possibly be null (even if you don't think it will be), it's best to add a check and specify what you want to happen in that null case.
+
+For example, we could add a null check to our view like this:
+```c#
+@model Movie
+
+@if (Model == null)
+{
+    <p>No movie details available.</p>
+}
+else
+{
+    <h1>Movie Details</h1>
+
+    <p>Title: @Model.Title</p>
+
+    <p>Genre: @Model.Genre</p>
+}
+```
+
+#### In The Controller
+
+Take a look at this action from the `MoviesController.cs`.
+
+```c#
+ // GET: /Movies/<id>
+[Route("Movies/{id:int}")]
+public IActionResult Show(int id)
+{
+    var movie = _context.Movies.Find(id);
+    return View(movie);
+}
+```
+
+<section class="call-to-action" markdown="1">
+With your partner: Discuss what might be null here? What might happen if those values are null?
+</section>
+
+Let's add some checks to make sure those values are not null.
+
+```c#
+[Route("Movies/{id:int}")]
+public IActionResult Show(int id)
+{
+    
+    if (id == null)

```

Comment:
This is an interesting question. I took this from the code the Entity Framework .NET docs use which looks like this. 

```
// GET: Students/Details/5
        public async Task<IActionResult> Details(int? id)
        {
            if (id == null)
            {
                return NotFound();
            }

            var student = await _context.Students
                .FirstOrDefaultAsync(m => m.ID == id);
            if (student == null)
            {
                return NotFound();
            }

            return View(student);
        }
```

But after looking closer they always have their parameters as optional. I'm not entirely sure why because they never do anything if the id isn't provided besides throwing a 404. Maybe better to get to this route and throw a 404 than to not match a route? And maybe in the future some nicer error handling would get added here?

I added a bit extra here to show only one null check needed if the parameter is required, and two if it is optional.